{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Parsing barcode variants from whole genome variant files\n",
    "\n",
    "Learn to subset variants from Zarr files that can be used in IDM models.\n",
    "\n",
    "The Zarr format is similar to HD5 that will allows to subset large files without loading into memory. For any whole genome sequencing file, variants in a population will be in a VCF format.\n",
    "\n",
    "Tutorials for guidance:\n",
    "\n",
    "http://alimanfoo.github.io/2016/06/10/scikit-allel-tour.html\n",
    "\n",
    "http://alimanfoo.github.io/2018/04/09/selecting-variants.html"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "# load packages - have been installed in vcf virtual env\n",
    "import os, sys\n",
    "import zarr\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "import allel  \n",
    "import dask.array as da\n",
    "from itertools import compress"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Functions"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "def read_zarr(chrom, zarr_path):\n",
    "    zarr_file = zarr_path + 'SNP_INDEL_Pf3D7_' + chrom + '_v3.zarr'\n",
    "    callset = zarr.open(zarr_file, mode='r')\n",
    "    return(callset)\n",
    "\n",
    "def gt_subset(callset, variant_bool, sample_bool):\n",
    "    gt_zarr = callset['calldata/GT']\n",
    "    gt_dask = allel.GenotypeDaskArray(gt_zarr)\n",
    "    gt_daskSub = gt_dask.subset(variant_bool, sample_bool).compute()\n",
    "    return(gt_daskSub)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "def overlap_avail(user_pos_list, callset):\n",
    "    zarr_index   = callset['variants/POS'][:].tolist()\n",
    "    user_overlap =  [i in user_pos_list for i in zarr_index]\n",
    "    return(user_overlap)\n",
    "    \n",
    "def overlap_filter(user_snps, filter_snps):\n",
    "    return(user_snps & filter_snps)\n",
    "\n",
    "def user_snps(user_df, chrom):\n",
    "    user_sub = user_df[user_df['chr'].str.contains(chrom)]\n",
    "    return(user_sub[\"position\"].tolist())\n",
    "    \n",
    "def variant_filter(callset, vsqlod_min, num_alt):\n",
    "    quality_set = callset['variants/FILTER_PASS'][:]\n",
    "    snp_set = callset['variants/is_snp'][:]\n",
    "    vsqlod_set = callset['variants/VQSLOD'][:]  > vsqlod_min\n",
    "    alt_set = callset['variants/numalt'][:] < num_alt + 1 \n",
    "    variant_hq = quality_set & snp_set & vsqlod_set & alt_set\n",
    "    return(variant_hq)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "def allele_freq(callset, variant_bool, sample_bool, num_alt):\n",
    "    gt_daskSub = gt_subset(callset, variant_bool, sample_bool)\n",
    "    # count number of alleles\n",
    "    ac = gt_daskSub.count_alleles(max_allele=num_alt)\n",
    "    sub_af = ac/ac.sum(axis=1, keepdims=True)\n",
    "    sub_df = pd.DataFrame(sub_af,\n",
    "                 index=[\"Pf3D7_\" + chrom + \"_v3:\" + str(x) for x in np.array(callset['variants/POS'][:])[variant_bool]],\n",
    "                 columns=[\"REF\"] + [\"ALT\" + str(x) for x in range(1, num_alt + 1)])\n",
    "    return(sub_df)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [],
   "source": [
    "def gt_merge(list1, list2, j):\n",
    "    dic = {-1:\"X\", \n",
    "           0: np.array(callset['variants/REF'][:])[variant_bool][j], \n",
    "           1: np.array(callset['variants/ALT'][:])[variant_bool][j,0], \n",
    "           2: np.array(callset['variants/ALT'][:])[variant_bool][j,1],\n",
    "           3: np.array(callset['variants/ALT'][:])[variant_bool][j,2]}\n",
    "    merged_list = []\n",
    "    for i in range(0, len(list1)):\n",
    "        if list1[i] == list2[i]:\n",
    "            if list1[i] == -1:\n",
    "                merged_list.append(\"X\")\n",
    "            else:\n",
    "                merged_list.append(list1[i])\n",
    "        else:\n",
    "            merged_list.append(\"N\") \n",
    "    merged_replace=[dic.get(n, n) for n in merged_list]  \n",
    "    return(merged_replace)\n",
    "\n",
    "def match_genotype(callset, variant_bool, sample_bool):\n",
    "    gt_daskSub = gt_subset(callset, variant_bool, sample_bool)\n",
    "    data =[]\n",
    "    for j in range(0, len(gt_daskSub)):\n",
    "        data.append(gt_merge(gt_daskSub[j,:,0], gt_daskSub[j,:,1], j))\n",
    "    gt_df = pd.DataFrame(data,\n",
    "                 index=[\"Pf3D7_\" + chrom + \"_v3:\" + str(x) for x in np.array(callset['variants/POS'][:])[variant_bool]],\n",
    "                 columns=list(compress(np.array(callset['samples']).tolist(), sample_bool)))\n",
    "    return(gt_df)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Run it together \n",
    "\n",
    "This section is specifically set up to run a for loop since Pf3k was divivde by chromosome. For other data sets, the above functions except reading in zarr object, should hold asumping it contains the same information in the Zarr file. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chr</th>\n",
       "      <th>position</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Pf3D7_01_v3</td>\n",
       "      <td>130339</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Pf3D7_01_v3</td>\n",
       "      <td>537322</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Pf3D7_02_v3</td>\n",
       "      <td>842805</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Pf3D7_04_v3</td>\n",
       "      <td>276127</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Pf3D7_05_v3</td>\n",
       "      <td>931606</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>Pf3D7_06_v3</td>\n",
       "      <td>145475</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>Pf3D7_06_v3</td>\n",
       "      <td>937752</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>Pf3D7_07_v3</td>\n",
       "      <td>221722</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>Pf3D7_07_v3</td>\n",
       "      <td>435497</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>Pf3D7_07_v3</td>\n",
       "      <td>489666</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>Pf3D7_07_v3</td>\n",
       "      <td>602559</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>Pf3D7_07_v3</td>\n",
       "      <td>616459</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>Pf3D7_07_v3</td>\n",
       "      <td>628392</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>Pf3D7_07_v3</td>\n",
       "      <td>736978</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>Pf3D7_07_v3</td>\n",
       "      <td>1359804</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>Pf3D7_08_v3</td>\n",
       "      <td>612596</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>Pf3D7_09_v3</td>\n",
       "      <td>634019</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>Pf3D7_10_v3</td>\n",
       "      <td>82375</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>Pf3D7_10_v3</td>\n",
       "      <td>1402510</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>Pf3D7_11_v3</td>\n",
       "      <td>119497</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>Pf3D7_11_v3</td>\n",
       "      <td>408600</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21</th>\n",
       "      <td>Pf3D7_13_v3</td>\n",
       "      <td>158412</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22</th>\n",
       "      <td>Pf3D7_13_v3</td>\n",
       "      <td>1429067</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23</th>\n",
       "      <td>Pf3D7_14_v3</td>\n",
       "      <td>755731</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "            chr  position\n",
       "0   Pf3D7_01_v3    130339\n",
       "1   Pf3D7_01_v3    537322\n",
       "2   Pf3D7_02_v3    842805\n",
       "3   Pf3D7_04_v3    276127\n",
       "4   Pf3D7_05_v3    931606\n",
       "5   Pf3D7_06_v3    145475\n",
       "6   Pf3D7_06_v3    937752\n",
       "7   Pf3D7_07_v3    221722\n",
       "8   Pf3D7_07_v3    435497\n",
       "9   Pf3D7_07_v3    489666\n",
       "10  Pf3D7_07_v3    602559\n",
       "11  Pf3D7_07_v3    616459\n",
       "12  Pf3D7_07_v3    628392\n",
       "13  Pf3D7_07_v3    736978\n",
       "14  Pf3D7_07_v3   1359804\n",
       "15  Pf3D7_08_v3    612596\n",
       "16  Pf3D7_09_v3    634019\n",
       "17  Pf3D7_10_v3     82375\n",
       "18  Pf3D7_10_v3   1402510\n",
       "19  Pf3D7_11_v3    119497\n",
       "20  Pf3D7_11_v3    408600\n",
       "21  Pf3D7_13_v3    158412\n",
       "22  Pf3D7_13_v3   1429067\n",
       "23  Pf3D7_14_v3    755731"
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# import the SNPs of interest - i.e. 24 barcode positions\n",
    "project_dir = '/home/jribado/Dropbox (IDM)/parasite_genetics/genomics/senegal'\n",
    "user_snp_file = os.path.join(project_dir, \"2008Daniels_BarcodePositions_Updated.txt\")\n",
    "user_snp_df = pd.read_csv(user_snp_file, sep='\\t', header=0)\n",
    "user_snp_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "['SenP008', 'SenP011', 'SenP019', 'SenP027']\n",
      "113\n"
     ]
    }
   ],
   "source": [
    "user_samples = '/home/jribado/Dropbox (IDM)/parasite_genetics/genomics/senegal/pf3k_bc_sampleOverlap.txt'\n",
    "with open(user_samples) as f:\n",
    "    sampleList = f.read().splitlines() \n",
    "print(sampleList[1:5])\n",
    "print(len(sampleList))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "                         REF      ALT1  ALT2  ALT3\n",
      "Pf3D7_01_v3:130339  0.302920  0.697080   0.0   0.0\n",
      "Pf3D7_01_v3:537322  0.021898  0.978102   0.0   0.0\n",
      "                   SenP005.02 SenP008.04 SenP011.02 SenP019.04 SenP027.02  \\\n",
      "Pf3D7_01_v3:130339          C          T          T          C          C   \n",
      "Pf3D7_01_v3:537322          A          A          A          A          A   \n",
      "\n",
      "                   SenP031.01 SenP051.02 SenP060.02 SenT001.08 SenT001.11  \\\n",
      "Pf3D7_01_v3:130339          T          C          T          T          N   \n",
      "Pf3D7_01_v3:537322          A          A          A          A          A   \n",
      "\n",
      "                    ... SenT230.08 SenT231.08 SenT232.08 SenT233.08  \\\n",
      "Pf3D7_01_v3:130339  ...          T          T          T          T   \n",
      "Pf3D7_01_v3:537322  ...          A          A          A          A   \n",
      "\n",
      "                   SenT235.08 SenT236.08 SenV034.04 SenV035.04 SenV042.05  \\\n",
      "Pf3D7_01_v3:130339          T          C          C          T          T   \n",
      "Pf3D7_01_v3:537322          A          A          A          A          A   \n",
      "\n",
      "                   SenV092.05  \n",
      "Pf3D7_01_v3:130339          T  \n",
      "Pf3D7_01_v3:537322          A  \n",
      "\n",
      "[2 rows x 137 columns]\n"
     ]
    }
   ],
   "source": [
    "project_dir = '/home/jribado/Dropbox (IDM)/parasite_genetics/genomics/senegal/'\n",
    "num_alt = 3\n",
    "genotypes, frequencies = [], []\n",
    "\n",
    "\n",
    "for i in range(1, 2):\n",
    "    chrom = \"{:02d}\".format(i)\n",
    "    # read in file\n",
    "    zarr_path = '/home/jribado/Dropbox (IDM)/Data, Dynamics, and Analytics Folder/Projects/malaria_pfcommunity/malaria_pf3k/pf3k_zarr/'\n",
    "    callset = read_zarr(chrom, zarr_path)\n",
    "    # identify variants in VCF that match user inputs\n",
    "    user_chrom_snps = user_snps(user_snp_df, chrom)\n",
    "    variant_bool = overlap_filter(overlap_avail(user_chrom_snps, callset), \n",
    "                                  variant_filter(callset, 2, num_alt))\n",
    "    # identify samples in VCF that match user inputs\n",
    "    sample_list = [x.split(\".\")[0] for x in np.array(callset['samples']).tolist()]\n",
    "    sample_bool = np.array([i in sampleList for i in sample_list]) \n",
    "    #sample_bool  = np.array([\"Sen\" in i for i in np.array(callset['samples']).tolist()])\n",
    "    freq = allele_freq(callset, variant_bool, sample_bool, 3)\n",
    "    frequencies.append(freq)\n",
    "    geno = match_genotype(callset, variant_bool, sample_bool)\n",
    "    genotypes.append(geno)\n",
    "\n",
    "appended_freq = pd.concat(frequencies)\n",
    "print(appended_freq)\n",
    "#appended_freq.to_csv('~/Desktop/Pf3k_wgsSenegalBCOverlapFrequencies24bc.txt') \n",
    "\n",
    "appended_geno = pd.concat(genotypes)\n",
    "print(appended_geno)\n",
    "#appended_geno.to_csv('~/Desktop/Pf3k_wgsSenegalBCOverlapsHaploidGenotype24bc.txt')  "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Incorporating depth and count data to improve called genotypes\n",
    "\n",
    "We are going to improve the genotype calls by incorporating minimum read counts.  "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 47,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<link rel=\"stylesheet\" href=\"//cdnjs.cloudflare.com/ajax/libs/jstree/3.3.3/themes/default/style.min.css\"/><div id=\"459efc1a-808a-47ad-85e6-aba280cfc335\" class=\"zarr-tree\"><ul><li data-jstree='{\"type\": \"Group\"}' class='jstree-open'><span>/</span><ul><li data-jstree='{\"type\": \"Group\"}' class=''><span>calldata</span><ul><li data-jstree='{\"type\": \"Array\"}' class=''><span>AD (159946, 2640, 4) int16</span></li><li data-jstree='{\"type\": \"Array\"}' class=''><span>DP (159946, 2640) int16</span></li><li data-jstree='{\"type\": \"Array\"}' class=''><span>GQ (159946, 2640) int8</span></li><li data-jstree='{\"type\": \"Array\"}' class=''><span>GT (159946, 2640, 2) int8</span></li><li data-jstree='{\"type\": \"Array\"}' class=''><span>MIN_DP (159946, 2640) int32</span></li><li data-jstree='{\"type\": \"Array\"}' class=''><span>PGT (159946, 2640) object</span></li><li data-jstree='{\"type\": \"Array\"}' class=''><span>PID (159946, 2640) object</span></li><li data-jstree='{\"type\": \"Array\"}' class=''><span>PL (159946, 2640, 3) int32</span></li><li data-jstree='{\"type\": \"Array\"}' class=''><span>RGQ (159946, 2640) int32</span></li><li data-jstree='{\"type\": \"Array\"}' class=''><span>SB (159946, 2640, 4) int32</span></li></ul></li><li data-jstree='{\"type\": \"Array\"}' class=''><span>samples (2640,) object</span></li><li data-jstree='{\"type\": \"Group\"}' class=''><span>variants</span><ul><li data-jstree='{\"type\": \"Array\"}' class=''><span>AC (159946, 3) int32</span></li><li data-jstree='{\"type\": \"Array\"}' class=''><span>AF (159946, 3) float32</span></li><li data-jstree='{\"type\": \"Array\"}' class=''><span>ALT (159946, 3) object</span></li><li data-jstree='{\"type\": \"Array\"}' class=''><span>AN (159946,) int32</span></li><li data-jstree='{\"type\": \"Array\"}' class=''><span>BaseQRankSum (159946,) float32</span></li><li data-jstree='{\"type\": \"Array\"}' class=''><span>CHROM (159946,) object</span></li><li data-jstree='{\"type\": \"Array\"}' class=''><span>ClippingRankSum (159946,) float32</span></li><li data-jstree='{\"type\": \"Array\"}' class=''><span>DP (159946,) int32</span></li><li data-jstree='{\"type\": \"Array\"}' class=''><span>DS (159946,) bool</span></li><li data-jstree='{\"type\": \"Array\"}' class=''><span>END (159946,) int32</span></li><li data-jstree='{\"type\": \"Array\"}' class=''><span>FILTER_Centromere (159946,) bool</span></li><li data-jstree='{\"type\": \"Array\"}' class=''><span>FILTER_InternalHypervariable (159946,) bool</span></li><li data-jstree='{\"type\": \"Array\"}' class=''><span>FILTER_LowQual (159946,) bool</span></li><li data-jstree='{\"type\": \"Array\"}' class=''><span>FILTER_Low_VQSLOD (159946,) bool</span></li><li data-jstree='{\"type\": \"Array\"}' class=''><span>FILTER_PASS (159946,) bool</span></li><li data-jstree='{\"type\": \"Array\"}' class=''><span>FILTER_SubtelomericHypervariable (159946,) bool</span></li><li data-jstree='{\"type\": \"Array\"}' class=''><span>FILTER_SubtelomericRepeat (159946,) bool</span></li><li data-jstree='{\"type\": \"Array\"}' class=''><span>FILTER_VQSRTrancheINDEL99.00to99.90 (159946,) bool</span></li><li data-jstree='{\"type\": \"Array\"}' class=''><span>FILTER_VQSRTrancheINDEL99.90to100.00 (159946,) bool</span></li><li data-jstree='{\"type\": \"Array\"}' class=''><span>FILTER_VQSRTrancheINDEL99.90to100.00+ (159946,) bool</span></li><li data-jstree='{\"type\": \"Array\"}' class=''><span>FILTER_VQSRTrancheSNP99.90to100.00 (159946,) bool</span></li><li data-jstree='{\"type\": \"Array\"}' class=''><span>FILTER_VQSRTrancheSNP99.90to100.00+ (159946,) bool</span></li><li data-jstree='{\"type\": \"Array\"}' class=''><span>FS (159946,) float32</span></li><li data-jstree='{\"type\": \"Array\"}' class=''><span>GC (159946,) float32</span></li><li data-jstree='{\"type\": \"Array\"}' class=''><span>HaplotypeScore (159946,) float32</span></li><li data-jstree='{\"type\": \"Array\"}' class=''><span>ID (159946,) object</span></li><li data-jstree='{\"type\": \"Array\"}' class=''><span>InbreedingCoeff (159946,) float32</span></li><li data-jstree='{\"type\": \"Array\"}' class=''><span>MLEAC (159946, 3) int32</span></li><li data-jstree='{\"type\": \"Array\"}' class=''><span>MLEAF (159946, 3) float32</span></li><li data-jstree='{\"type\": \"Array\"}' class=''><span>MQ (159946,) float32</span></li><li data-jstree='{\"type\": \"Array\"}' class=''><span>MQRankSum (159946,) float32</span></li><li data-jstree='{\"type\": \"Array\"}' class=''><span>NEGATIVE_TRAIN_SITE (159946,) bool</span></li><li data-jstree='{\"type\": \"Array\"}' class=''><span>POS (159946,) int32</span></li><li data-jstree='{\"type\": \"Array\"}' class=''><span>POSITIVE_TRAIN_SITE (159946,) bool</span></li><li data-jstree='{\"type\": \"Array\"}' class=''><span>QD (159946,) float32</span></li><li data-jstree='{\"type\": \"Array\"}' class=''><span>QUAL (159946,) float32</span></li><li data-jstree='{\"type\": \"Array\"}' class=''><span>REF (159946,) object</span></li><li data-jstree='{\"type\": \"Array\"}' class=''><span>RPA (159946,) int32</span></li><li data-jstree='{\"type\": \"Array\"}' class=''><span>RU (159946,) object</span></li><li data-jstree='{\"type\": \"Array\"}' class=''><span>ReadPosRankSum (159946,) float32</span></li><li data-jstree='{\"type\": \"Array\"}' class=''><span>RegionType (159946,) object</span></li><li data-jstree='{\"type\": \"Array\"}' class=''><span>SNPEFF_AMINO_ACID_CHANGE (159946,) object</span></li><li data-jstree='{\"type\": \"Array\"}' class=''><span>SNPEFF_CODON_CHANGE (159946,) object</span></li><li data-jstree='{\"type\": \"Array\"}' class=''><span>SNPEFF_EFFECT (159946,) object</span></li><li data-jstree='{\"type\": \"Array\"}' class=''><span>SNPEFF_EXON_ID (159946,) object</span></li><li data-jstree='{\"type\": \"Array\"}' class=''><span>SNPEFF_FUNCTIONAL_CLASS (159946,) object</span></li><li data-jstree='{\"type\": \"Array\"}' class=''><span>SNPEFF_GENE_BIOTYPE (159946,) object</span></li><li data-jstree='{\"type\": \"Array\"}' class=''><span>SNPEFF_GENE_NAME (159946,) object</span></li><li data-jstree='{\"type\": \"Array\"}' class=''><span>SNPEFF_IMPACT (159946,) object</span></li><li data-jstree='{\"type\": \"Array\"}' class=''><span>SNPEFF_TRANSCRIPT_ID (159946,) object</span></li><li data-jstree='{\"type\": \"Array\"}' class=''><span>SOR (159946,) float32</span></li><li data-jstree='{\"type\": \"Array\"}' class=''><span>STR (159946,) bool</span></li><li data-jstree='{\"type\": \"Array\"}' class=''><span>VQSLOD (159946,) float32</span></li><li data-jstree='{\"type\": \"Array\"}' class=''><span>VariantType (159946,) object</span></li><li data-jstree='{\"type\": \"Array\"}' class=''><span>altlen (159946, 3) int32</span></li><li data-jstree='{\"type\": \"Array\"}' class=''><span>culprit (159946,) object</span></li><li data-jstree='{\"type\": \"Array\"}' class=''><span>is_snp (159946,) bool</span></li><li data-jstree='{\"type\": \"Array\"}' class=''><span>numalt (159946,) int32</span></li><li data-jstree='{\"type\": \"Array\"}' class=''><span>set (159946,) object</span></li></ul></li></ul></li></ul></div>\n",
       "<script>\n",
       "    if (!require.defined('jquery')) {\n",
       "        require.config({\n",
       "            paths: {\n",
       "                jquery: '//cdnjs.cloudflare.com/ajax/libs/jquery/1.12.1/jquery.min'\n",
       "            },\n",
       "        });\n",
       "    }\n",
       "    if (!require.defined('jstree')) {\n",
       "        require.config({\n",
       "            paths: {\n",
       "                jstree: '//cdnjs.cloudflare.com/ajax/libs/jstree/3.3.3/jstree.min'\n",
       "            },\n",
       "        });\n",
       "    }\n",
       "    require(['jstree'], function() {\n",
       "        $('#459efc1a-808a-47ad-85e6-aba280cfc335').jstree({\n",
       "            types: {\n",
       "                Group: {\n",
       "                    icon: \"fa fa-folder\"\n",
       "                },\n",
       "                Array: {\n",
       "                    icon: \"fa fa-table\"\n",
       "                }\n",
       "            },\n",
       "            plugins: [\"types\"]\n",
       "        });\n",
       "    });\n",
       "</script>\n"
      ],
      "text/plain": [
       "/\n",
       " ├── calldata\n",
       " │   ├── AD (159946, 2640, 4) int16\n",
       " │   ├── DP (159946, 2640) int16\n",
       " │   ├── GQ (159946, 2640) int8\n",
       " │   ├── GT (159946, 2640, 2) int8\n",
       " │   ├── MIN_DP (159946, 2640) int32\n",
       " │   ├── PGT (159946, 2640) object\n",
       " │   ├── PID (159946, 2640) object\n",
       " │   ├── PL (159946, 2640, 3) int32\n",
       " │   ├── RGQ (159946, 2640) int32\n",
       " │   └── SB (159946, 2640, 4) int32\n",
       " ├── samples (2640,) object\n",
       " └── variants\n",
       "     ├── AC (159946, 3) int32\n",
       "     ├── AF (159946, 3) float32\n",
       "     ├── ALT (159946, 3) object\n",
       "     ├── AN (159946,) int32\n",
       "     ├── BaseQRankSum (159946,) float32\n",
       "     ├── CHROM (159946,) object\n",
       "     ├── ClippingRankSum (159946,) float32\n",
       "     ├── DP (159946,) int32\n",
       "     ├── DS (159946,) bool\n",
       "     ├── END (159946,) int32\n",
       "     ├── FILTER_Centromere (159946,) bool\n",
       "     ├── FILTER_InternalHypervariable (159946,) bool\n",
       "     ├── FILTER_LowQual (159946,) bool\n",
       "     ├── FILTER_Low_VQSLOD (159946,) bool\n",
       "     ├── FILTER_PASS (159946,) bool\n",
       "     ├── FILTER_SubtelomericHypervariable (159946,) bool\n",
       "     ├── FILTER_SubtelomericRepeat (159946,) bool\n",
       "     ├── FILTER_VQSRTrancheINDEL99.00to99.90 (159946,) bool\n",
       "     ├── FILTER_VQSRTrancheINDEL99.90to100.00 (159946,) bool\n",
       "     ├── FILTER_VQSRTrancheINDEL99.90to100.00+ (159946,) bool\n",
       "     ├── FILTER_VQSRTrancheSNP99.90to100.00 (159946,) bool\n",
       "     ├── FILTER_VQSRTrancheSNP99.90to100.00+ (159946,) bool\n",
       "     ├── FS (159946,) float32\n",
       "     ├── GC (159946,) float32\n",
       "     ├── HaplotypeScore (159946,) float32\n",
       "     ├── ID (159946,) object\n",
       "     ├── InbreedingCoeff (159946,) float32\n",
       "     ├── MLEAC (159946, 3) int32\n",
       "     ├── MLEAF (159946, 3) float32\n",
       "     ├── MQ (159946,) float32\n",
       "     ├── MQRankSum (159946,) float32\n",
       "     ├── NEGATIVE_TRAIN_SITE (159946,) bool\n",
       "     ├── POS (159946,) int32\n",
       "     ├── POSITIVE_TRAIN_SITE (159946,) bool\n",
       "     ├── QD (159946,) float32\n",
       "     ├── QUAL (159946,) float32\n",
       "     ├── REF (159946,) object\n",
       "     ├── RPA (159946,) int32\n",
       "     ├── RU (159946,) object\n",
       "     ├── ReadPosRankSum (159946,) float32\n",
       "     ├── RegionType (159946,) object\n",
       "     ├── SNPEFF_AMINO_ACID_CHANGE (159946,) object\n",
       "     ├── SNPEFF_CODON_CHANGE (159946,) object\n",
       "     ├── SNPEFF_EFFECT (159946,) object\n",
       "     ├── SNPEFF_EXON_ID (159946,) object\n",
       "     ├── SNPEFF_FUNCTIONAL_CLASS (159946,) object\n",
       "     ├── SNPEFF_GENE_BIOTYPE (159946,) object\n",
       "     ├── SNPEFF_GENE_NAME (159946,) object\n",
       "     ├── SNPEFF_IMPACT (159946,) object\n",
       "     ├── SNPEFF_TRANSCRIPT_ID (159946,) object\n",
       "     ├── SOR (159946,) float32\n",
       "     ├── STR (159946,) bool\n",
       "     ├── VQSLOD (159946,) float32\n",
       "     ├── VariantType (159946,) object\n",
       "     ├── altlen (159946, 3) int32\n",
       "     ├── culprit (159946,) object\n",
       "     ├── is_snp (159946,) bool\n",
       "     ├── numalt (159946,) int32\n",
       "     └── set (159946,) object"
      ]
     },
     "execution_count": 47,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "chrom=\"01\"\n",
    "zarr_path = '/home/jribado/Dropbox (IDM)/Data, Dynamics, and Analytics Folder/Projects/malaria_pfcommunity/malaria_pf3k/pf3k_zarr/'\n",
    "callset = read_zarr(chrom, zarr_path)\n",
    "callset.tree()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [],
   "source": [
    "user_chrom_snps = user_snps(user_snp_df, chrom)\n",
    "variant_bool = overlap_filter(overlap_avail(user_chrom_snps, callset), variant_filter(callset, 2, 2))\n",
    "    # identify samples in VCF that match user inputs\n",
    "sample_list = [x.split(\".\")[0] for x in np.array(callset['samples']).tolist()]\n",
    "sample_bool = np.array([i in sampleList for i in sample_list])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 48,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div class=\"allel allel-DisplayAs2D\"><span>&lt;GenotypeArray shape=(2, 137, 2) dtype=int8&gt;</span><table><thead><tr><th></th><th style=\"text-align: center\">0</th><th style=\"text-align: center\">1</th><th style=\"text-align: center\">2</th><th style=\"text-align: center\">3</th><th style=\"text-align: center\">4</th><th style=\"text-align: center\">...</th><th style=\"text-align: center\">132</th><th style=\"text-align: center\">133</th><th style=\"text-align: center\">134</th><th style=\"text-align: center\">135</th><th style=\"text-align: center\">136</th></tr></thead><tbody><tr><th style=\"text-align: center; background-color: white; border-right: 1px solid black; \">0</th><td style=\"text-align: center\">0/0</td><td style=\"text-align: center\">1/1</td><td style=\"text-align: center\">1/1</td><td style=\"text-align: center\">0/0</td><td style=\"text-align: center\">0/0</td><td style=\"text-align: center\">...</td><td style=\"text-align: center\">0/0</td><td style=\"text-align: center\">0/0</td><td style=\"text-align: center\">1/1</td><td style=\"text-align: center\">1/1</td><td style=\"text-align: center\">1/1</td></tr><tr><th style=\"text-align: center; background-color: white; border-right: 1px solid black; \">1</th><td style=\"text-align: center\">1/1</td><td style=\"text-align: center\">1/1</td><td style=\"text-align: center\">1/1</td><td style=\"text-align: center\">1/1</td><td style=\"text-align: center\">1/1</td><td style=\"text-align: center\">...</td><td style=\"text-align: center\">1/1</td><td style=\"text-align: center\">1/1</td><td style=\"text-align: center\">1/1</td><td style=\"text-align: center\">1/1</td><td style=\"text-align: center\">1/1</td></tr></tbody></table></div>"
      ],
      "text/plain": [
       "<GenotypeArray shape=(2, 137, 2) dtype=int8>\n",
       "0/0 1/1 1/1 0/0 0/0 ... 0/0 0/0 1/1 1/1 1/1\n",
       "1/1 1/1 1/1 1/1 1/1 ... 1/1 1/1 1/1 1/1 1/1"
      ]
     },
     "execution_count": 48,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "gt_zarr = callset['calldata/GT']\n",
    "gt_dask = allel.GenotypeDaskArray(gt_zarr)\n",
    "gt_daskSub = gt_dask.subset(variant_bool, sample_bool).compute()\n",
    "gt_daskSub"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 49,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[[ 38  89  23  58  47  47  44  40 135  74 119 270 108 174 213  44  66  35\n",
      "   35 309  25  82 242 273 101  70  58  84  45 352 197  37  42  18 110 112\n",
      "  105  45 239 130  32  55  29  65  63  37  35  20 155  58 125 174  33 141\n",
      "  234 118 213  18 131 162 185  70 221  27 165  36 258 534 223  50  88 128\n",
      "   56 107 122 263 423 122  23  40  40  87  52 402 199 194  92  35  72 383\n",
      "   60  67 148 543 179 118  43  44  36  35  42 221 148 282  36  48  87 269\n",
      "  130 239 284 315 166  41 191 235 274 197  36  36 130  41 145  76  41  36\n",
      "  115 103  98  94 159 153  36  36  49  84 105]\n",
      " [ 73  96  66  78 146  67 154  67 121  66 163 411 109 216 253  43  91  32\n",
      "  130 387  23  86 299 355 116 535  76 650  26 406 249  37 287  26 117 181\n",
      "  147 107 518 138  34  53  45  53 355 233 140  21 392  69  92 201  52 153\n",
      "  282 204 384  11 193 161 197 257 219  25 249  32 327 695 235  49  81 113\n",
      "  514 144  93 208 302  77  19 150  81  83 372 690 199 248 113 127  83 402\n",
      "  545 112 178 594 124 793 128 165  67 156 535 357 233 347  37  37  93 339\n",
      "  153 340 353 488 151 120 197 350 346 154 135  52 155  42 211  64  95  97\n",
      "   92 172 122 161 199  73 100  75  71 148 132]]\n"
     ]
    }
   ],
   "source": [
    "# pull total number of high quality reads that align to each position per sample\n",
    "dp_zarr = callset['calldata/DP']\n",
    "dp_dask = allel.AlleleCountsDaskArray(dp_zarr)\n",
    "dp_variant_selection = dp_dask.compress(variant_bool, axis=0).compute()\n",
    "dp_variant_selection = dp_variant_selection[:, sample_bool]\n",
    "print(dp_variant_selection)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 50,
   "metadata": {},
   "outputs": [
    {
     "ename": "ValueError",
     "evalue": "either both or neither of x and y should be given",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mValueError\u001b[0m                                Traceback (most recent call last)",
      "\u001b[0;32m<ipython-input-50-6a7dadb7da10>\u001b[0m in \u001b[0;36m<module>\u001b[0;34m\u001b[0m\n\u001b[1;32m      1\u001b[0m \u001b[0mcutoff\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;36m100\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m----> 2\u001b[0;31m \u001b[0mdp_lt_cutoff\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mnp\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mwhere\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mdp_variant_selection\u001b[0m\u001b[0;34m<\u001b[0m\u001b[0mcutoff\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0;36m0\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m      3\u001b[0m \u001b[0mdp_lt_cutoff\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m<__array_function__ internals>\u001b[0m in \u001b[0;36mwhere\u001b[0;34m(*args, **kwargs)\u001b[0m\n",
      "\u001b[0;31mValueError\u001b[0m: either both or neither of x and y should be given"
     ]
    }
   ],
   "source": [
    "cutoff=100\n",
    "dp_lt_cutoff=np.where(dp_variant_selection<cutoff,0)\n",
    "dp_lt_cutoff"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "metadata": {},
   "outputs": [],
   "source": [
    "# get the number of reads that align to each of the alleles\n",
    "ad_zarr=callset['calldata/AD']\n",
    "ad_array=da.from_zarr(ad_zarr)\n",
    "ad_variant_selection=ad_array[variant_bool]\n",
    "ad_variant_selection=ad_variant_selection[:,sample_bool,:]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 67,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[130339, 537322]"
      ]
     },
     "execution_count": 67,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "user_chrom_snps"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "dp_dask = allel.AlleleCountsDaskArray(dp_zarr)\n",
    "\n",
    "user_chrom_snps = user_snps(user_snp_df, chrom)\n",
    "variant_bool = overlap_filter(overlap_avail(user_chrom_snps, callset), variant_filter(callset, 2, 2))\n",
    "dp_variant_selection = dp_dask.compress(variant_bool, axis=0).compute()\n",
    "dp_variant_selection = dp_variant_selection[:, sample_bool]\n",
    "\n",
    "def overlap_avail(user_pos_list, callset):\n",
    "    zarr_index   = callset['variants/POS'][:].tolist()\n",
    "    user_overlap =  [i in user_pos_list for i in zarr_index]\n",
    "    return(user_overlap)\n",
    "\n",
    "class SNP:\n",
    "    def __init__(self, chromosome, position):\n",
    "        self.chromosome = chromosome\n",
    "        self.position = position\n",
    "        variant_bool = overlap_filter(overlap_avail(position, callset), variant_filter(callset, 2, 2)\n",
    "        \n",
    "        continue\n",
    "                                      \n",
    "\n",
    "    \n",
    "        "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 57,
   "metadata": {},
   "outputs": [],
   "source": [
    "locus1_totalreads = dp_variant_selection[0]\n",
    "locus2_totalreads = dp_variant_selection[1]\n",
    "\n",
    "locus1_ad_array_ref = ad_variant_selection[0,:,0]\n",
    "locus2_ad_array_ref = ad_variant_selection[1,:,0]\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 66,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[ 38   2   0  58  47   0  44   0   0  48   0   0   0  31   1   0   0   0\n",
      "  35   0   0   0   1   0   0  70   1  84   0   0   0  37  42   0   0   0\n",
      "   0  45   0   0   0   0   0   0  63  37  35   0   0   0   0   0  33   0\n",
      "   0   0   0   0   0 162   0  70   0   0   0  36   0  12   0   1   0   0\n",
      "  56   0  17   0   0   0  23  40  40   0  52   1   0   0   0  35   0   0\n",
      "  60   0   0   1   1 118  43  44  36  35  42   1   0 247  36   0   0  48\n",
      "   0   0  25 315   0  41   2   1   1   6  36  36   0   0   0   0  41  36\n",
      "   0   0   0   0   1   1  36  36   0   0   0]\n",
      "[ 38  89  23  58  47  47  44  40 135  74 119 270 108 174 213  44  66  35\n",
      "  35 309  25  82 242 273 101  70  58  84  45 352 197  37  42  18 110 112\n",
      " 105  45 239 130  32  55  29  65  63  37  35  20 155  58 125 174  33 141\n",
      " 234 118 213  18 131 162 185  70 221  27 165  36 258 534 223  50  88 128\n",
      "  56 107 122 263 423 122  23  40  40  87  52 402 199 194  92  35  72 383\n",
      "  60  67 148 543 179 118  43  44  36  35  42 221 148 282  36  48  87 269\n",
      " 130 239 284 315 166  41 191 235 274 197  36  36 130  41 145  76  41  36\n",
      " 115 103  98  94 159 153  36  36  49  84 105]\n",
      "[0.         0.97752809 1.         0.         0.         1.\n",
      " 0.         1.         1.         0.35135135 1.         1.\n",
      " 1.         0.82183908 0.99530516 1.         1.         1.\n",
      " 0.         1.         1.         1.         0.99586777 1.\n",
      " 1.         0.         0.98275862 0.         1.         1.\n",
      " 1.         0.         0.         1.         1.         1.\n",
      " 1.         0.         1.         1.         1.         1.\n",
      " 1.         1.         0.         0.         0.         1.\n",
      " 1.         1.         1.         1.         0.         1.\n",
      " 1.         1.         1.         1.         1.         0.\n",
      " 1.         0.         1.         1.         1.         0.\n",
      " 1.         0.97752809 1.         0.98       1.         1.\n",
      " 0.         1.         0.86065574 1.         1.         1.\n",
      " 0.         0.         0.         1.         0.         0.99751244\n",
      " 1.         1.         1.         0.         1.         1.\n",
      " 0.         1.         1.         0.99815838 0.99441341 0.\n",
      " 0.         0.         0.         0.         0.         0.99547511\n",
      " 1.         0.12411348 0.         1.         1.         0.82156134\n",
      " 1.         1.         0.91197183 0.         1.         0.\n",
      " 0.9895288  0.99574468 0.99635036 0.96954315 0.         0.\n",
      " 1.         1.         1.         1.         0.         0.\n",
      " 1.         1.         1.         1.         0.99371069 0.99346405\n",
      " 0.         0.         1.         1.         1.        ]\n"
     ]
    }
   ],
   "source": [
    "print(np.asarray(locus1_ad_array_ref))\n",
    "print(locus1_totalreads )\n",
    "\n",
    "locus1_refprop = (locus1_totalreads - locus1_ad_array_ref ) / locus1_totalreads\n",
    "locus1_alt1prop = (locus1_totalreads = locus1_ada)\n",
    "print(np.asarray(locus1_refprop))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 55,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<table>\n",
       "<tr>\n",
       "<td>\n",
       "<table>\n",
       "  <thead>\n",
       "    <tr><td> </td><th> Array </th><th> Chunk </th></tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr><th> Bytes </th><td> 548 B </td> <td> 128 B </td></tr>\n",
       "    <tr><th> Shape </th><td> (2, 137) </td> <td> (1, 64) </td></tr>\n",
       "    <tr><th> Count </th><td> 223 Tasks </td><td> 6 Chunks </td></tr>\n",
       "    <tr><th> Type </th><td> int16 </td><td> numpy.ndarray </td></tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</td>\n",
       "<td>\n",
       "<svg width=\"170\" height=\"77\" style=\"stroke:rgb(0,0,0);stroke-width:1\" >\n",
       "\n",
       "  <!-- Horizontal lines -->\n",
       "  <line x1=\"0\" y1=\"0\" x2=\"120\" y2=\"0\" style=\"stroke-width:2\" />\n",
       "  <line x1=\"0\" y1=\"13\" x2=\"120\" y2=\"13\" />\n",
       "  <line x1=\"0\" y1=\"27\" x2=\"120\" y2=\"27\" style=\"stroke-width:2\" />\n",
       "\n",
       "  <!-- Vertical lines -->\n",
       "  <line x1=\"0\" y1=\"0\" x2=\"0\" y2=\"27\" style=\"stroke-width:2\" />\n",
       "  <line x1=\"49\" y1=\"0\" x2=\"49\" y2=\"27\" />\n",
       "  <line x1=\"105\" y1=\"0\" x2=\"105\" y2=\"27\" />\n",
       "  <line x1=\"120\" y1=\"0\" x2=\"120\" y2=\"27\" style=\"stroke-width:2\" />\n",
       "\n",
       "  <!-- Colored Rectangle -->\n",
       "  <polygon points=\"0.000000,0.000000 120.000000,0.000000 120.000000,27.314855 0.000000,27.314855\" style=\"fill:#ECB172A0;stroke-width:0\"/>\n",
       "\n",
       "  <!-- Text -->\n",
       "  <text x=\"60.000000\" y=\"47.314855\" font-size=\"1.0rem\" font-weight=\"100\" text-anchor=\"middle\" >137</text>\n",
       "  <text x=\"140.000000\" y=\"13.657428\" font-size=\"1.0rem\" font-weight=\"100\" text-anchor=\"middle\" transform=\"rotate(0,140.000000,13.657428)\">2</text>\n",
       "</svg>\n",
       "</td>\n",
       "</tr>\n",
       "</table>"
      ],
      "text/plain": [
       "dask.array<getitem, shape=(2, 137), dtype=int16, chunksize=(1, 64), chunktype=numpy.ndarray>"
      ]
     },
     "execution_count": 55,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ad_array_ref"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 51,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[[ 38   2   0  58  47   0  44   0   0  48   0   0   0  31   1   0   0   0\n",
      "   35   0   0   0   1   0   0  70   1  84   0   0   0  37  42   0   0   0\n",
      "    0  45   0   0   0   0   0   0  63  37  35   0   0   0   0   0  33   0\n",
      "    0   0   0   0   0 162   0  70   0   0   0  36   0  12   0   1   0   0\n",
      "   56   0  17   0   0   0  23  40  40   0  52   1   0   0   0  35   0   0\n",
      "   60   0   0   1   1 118  43  44  36  35  42   1   0 247  36   0   0  48\n",
      "    0   0  25 315   0  41   2   1   1   6  36  36   0   0   0   0  41  36\n",
      "    0   0   0   0   1   1  36  36   0   0   0]\n",
      " [  0   2   0   1   0   0   0   0   0   0   0   0  27 179   0   0   0   0\n",
      "    0   0   0   0   0   0   0   0   0   1   0   0   0   0   0   0   0   0\n",
      "    0   0   3   0  34   0   0   0   1   0   0   0 339   0   0   0   0   0\n",
      "    0   0  41   0   0   0   0   0   0   0   0   0   0   0   0   0   0   1\n",
      "    0   0   0   0   0   0   0   0   0   0   2   1   0   0   0   0   0   0\n",
      "    0   0   0   2   0   2   0   0   0   0   0   0   0   0   0   0   0   0\n",
      "    0   0   1   1   0   0   0   0   1   0   0   0   0   0  46   0   0   0\n",
      "    0   0   0   0   0   0   0   0   0   0   0]]\n",
      "[[1.         0.02247191 0.         1.         1.         0.\n",
      "  1.         0.         0.         0.64864865 0.         0.\n",
      "  0.         0.17816092 0.00469484 0.         0.         0.\n",
      "  1.         0.         0.         0.         0.00413223 0.\n",
      "  0.         1.         0.01724138 1.         0.         0.\n",
      "  0.         1.         1.         0.         0.         0.\n",
      "  0.         1.         0.         0.         0.         0.\n",
      "  0.         0.         1.         1.         1.         0.\n",
      "  0.         0.         0.         0.         1.         0.\n",
      "  0.         0.         0.         0.         0.         1.\n",
      "  0.         1.         0.         0.         0.         1.\n",
      "  0.         0.02247191 0.         0.02       0.         0.\n",
      "  1.         0.         0.13934426 0.         0.         0.\n",
      "  1.         1.         1.         0.         1.         0.00248756\n",
      "  0.         0.         0.         1.         0.         0.\n",
      "  1.         0.         0.         0.00184162 0.00558659 1.\n",
      "  1.         1.         1.         1.         1.         0.00452489\n",
      "  0.         0.87588652 1.         0.         0.         0.17843866\n",
      "  0.         0.         0.08802817 1.         0.         1.\n",
      "  0.0104712  0.00425532 0.00364964 0.03045685 1.         1.\n",
      "  0.         0.         0.         0.         1.         1.\n",
      "  0.         0.         0.         0.         0.00628931 0.00653595\n",
      "  1.         1.         0.         0.         0.        ]\n",
      " [0.         0.02083333 0.         0.01282051 0.         0.\n",
      "  0.         0.         0.         0.         0.         0.\n",
      "  0.24770642 0.8287037  0.         0.         0.         0.\n",
      "  0.         0.         0.         0.         0.         0.\n",
      "  0.         0.         0.         0.00153846 0.         0.\n",
      "  0.         0.         0.         0.         0.         0.\n",
      "  0.         0.         0.00579151 0.         1.         0.\n",
      "  0.         0.         0.0028169  0.         0.         0.\n",
      "  0.86479592 0.         0.         0.         0.         0.\n",
      "  0.         0.         0.10677083 0.         0.         0.\n",
      "  0.         0.         0.         0.         0.         0.\n",
      "  0.         0.         0.         0.         0.         0.00884956\n",
      "  0.         0.         0.         0.         0.         0.\n",
      "  0.         0.         0.         0.         0.00537634 0.00144928\n",
      "  0.         0.         0.         0.         0.         0.\n",
      "  0.         0.         0.         0.003367   0.         0.00252207\n",
      "  0.         0.         0.         0.         0.         0.\n",
      "  0.         0.         0.         0.         0.         0.\n",
      "  0.         0.         0.00283286 0.00204918 0.         0.\n",
      "  0.         0.         0.00289017 0.         0.         0.\n",
      "  0.         0.         0.21800948 0.         0.         0.\n",
      "  0.         0.         0.         0.         0.         0.\n",
      "  0.         0.         0.         0.         0.        ]]\n"
     ]
    }
   ],
   "source": [
    "# pull the first item in the allele count array to get the reference count\n",
    "ad_array_ref=ad_variant_selection[:,:,0]\n",
    "ad_array_ref=ad_array_ref.compute()\n",
    "print(ad_array_ref)\n",
    "\n",
    "# ignore zero division for site where one population may have an allele but other samples may not have adequate coverage\n",
    "np.seterr(divide='ignore', invalid='ignore')\n",
    "a=ad_array_ref/dp_variant_selection\n",
    "print(a)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 45,
   "metadata": {},
   "outputs": [],
   "source": [
    "a[a<0.1] = 1\n",
    "a[a>=0.9] = 0\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 46,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array([[0.        , 0.        , 0.        , 0.        , 0.        ,\n",
       "        0.        , 0.        , 0.        , 0.        , 0.64864865,\n",
       "        0.        , 0.        , 0.        , 0.17816092, 0.        ,\n",
       "        0.        , 0.        , 0.        , 0.        , 0.        ,\n",
       "        0.        , 0.        , 0.        , 0.        , 0.        ,\n",
       "        0.        , 0.        , 0.        , 0.        , 0.        ,\n",
       "        0.        , 0.        , 0.        , 0.        , 0.        ,\n",
       "        0.        , 0.        , 0.        , 0.        , 0.        ,\n",
       "        0.        , 0.        , 0.        , 0.        , 0.        ,\n",
       "        0.        , 0.        , 0.        , 0.        , 0.        ,\n",
       "        0.        , 0.        , 0.        , 0.        , 0.        ,\n",
       "        0.        , 0.        , 0.        , 0.        , 0.        ,\n",
       "        0.        , 0.        , 0.        , 0.        , 0.        ,\n",
       "        0.        , 0.        , 0.        , 0.        , 0.        ,\n",
       "        0.        , 0.        , 0.        , 0.        , 0.13934426,\n",
       "        0.        , 0.        , 0.        , 0.        , 0.        ,\n",
       "        0.        , 0.        , 0.        , 0.        , 0.        ,\n",
       "        0.        , 0.        , 0.        , 0.        , 0.        ,\n",
       "        0.        , 0.        , 0.        , 0.        , 0.        ,\n",
       "        0.        , 0.        , 0.        , 0.        , 0.        ,\n",
       "        0.        , 0.        , 0.        , 0.87588652, 0.        ,\n",
       "        0.        , 0.        , 0.17843866, 0.        , 0.        ,\n",
       "        0.        , 0.        , 0.        , 0.        , 0.        ,\n",
       "        0.        , 0.        , 0.        , 0.        , 0.        ,\n",
       "        0.        , 0.        , 0.        , 0.        , 0.        ,\n",
       "        0.        , 0.        , 0.        , 0.        , 0.        ,\n",
       "        0.        , 0.        , 0.        , 0.        , 0.        ,\n",
       "        0.        , 0.        ],\n",
       "       [0.        , 0.        , 0.        , 0.        , 0.        ,\n",
       "        0.        , 0.        , 0.        , 0.        , 0.        ,\n",
       "        0.        , 0.        , 0.24770642, 0.8287037 , 0.        ,\n",
       "        0.        , 0.        , 0.        , 0.        , 0.        ,\n",
       "        0.        , 0.        , 0.        , 0.        , 0.        ,\n",
       "        0.        , 0.        , 0.        , 0.        , 0.        ,\n",
       "        0.        , 0.        , 0.        , 0.        , 0.        ,\n",
       "        0.        , 0.        , 0.        , 0.        , 0.        ,\n",
       "        0.        , 0.        , 0.        , 0.        , 0.        ,\n",
       "        0.        , 0.        , 0.        , 0.86479592, 0.        ,\n",
       "        0.        , 0.        , 0.        , 0.        , 0.        ,\n",
       "        0.        , 0.10677083, 0.        , 0.        , 0.        ,\n",
       "        0.        , 0.        , 0.        , 0.        , 0.        ,\n",
       "        0.        , 0.        , 0.        , 0.        , 0.        ,\n",
       "        0.        , 0.        , 0.        , 0.        , 0.        ,\n",
       "        0.        , 0.        , 0.        , 0.        , 0.        ,\n",
       "        0.        , 0.        , 0.        , 0.        , 0.        ,\n",
       "        0.        , 0.        , 0.        , 0.        , 0.        ,\n",
       "        0.        , 0.        , 0.        , 0.        , 0.        ,\n",
       "        0.        , 0.        , 0.        , 0.        , 0.        ,\n",
       "        0.        , 0.        , 0.        , 0.        , 0.        ,\n",
       "        0.        , 0.        , 0.        , 0.        , 0.        ,\n",
       "        0.        , 0.        , 0.        , 0.        , 0.        ,\n",
       "        0.        , 0.        , 0.        , 0.        , 0.        ,\n",
       "        0.        , 0.        , 0.21800948, 0.        , 0.        ,\n",
       "        0.        , 0.        , 0.        , 0.        , 0.        ,\n",
       "        0.        , 0.        , 0.        , 0.        , 0.        ,\n",
       "        0.        , 0.        ]])"
      ]
     },
     "execution_count": 46,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "a"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.9"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
